examples/lock-app/silabs/src/LockManager.cpp

```diff
@@ -743,25 +923,78 @@ bool LockManager::setLockState(chip::EndpointId endpointId, const Nullable<chip:
-    for (const auto & currentCredential : mLockCredentials[to_underlying(CredentialTypeEnum::kPin)])
+    // Check all pin codes associated to all users to see if this pin code exists
+    for (int userIndex = 0; userIndex < kMaxUsers; userIndex++)
     {
+        // Get user data to obtain currentCredentialCount
+        chip::StorageKeyName userKey = LockUserEndpoint(userIndex, endpointId);
+
+        uint16_t size = static_cast<uint16_t>(sizeof(LockUserInfo));
+
+        error = mStorage->SyncGetKeyValue(userKey.KeyName(), &mUserInStorage, size);
 
-        if (currentCredential.status == DlCredentialStatus::kAvailable)
+        // No user exists at this index
+        if (error != CHIP_NO_ERROR)
         {
             continue;
         }
 
-        if (currentCredential.credentialData.data_equal(pin.Value()))
+        chip::StorageKeyName credentialKey = LockUserCredentialMap(userIndex);
+
+        uint16_t credentialStructSize = static_cast<uint16_t>(sizeof(CredentialStruct) * mUserInStorage.currentCredentialCount);
+
+        // Get array of credential indices and types associated to user
+        error = mStorage->SyncGetKeyValue(credentialKey.KeyName(), &mCredentials, credentialStructSize);
+
+        // No credential data associated with user
+        if (error != CHIP_NO_ERROR)
         {
-            ChipLogDetail(Zcl,
-                          "Lock App: specified PIN code was found in the database, setting lock state to \"%s\" [endpointId=%d]",
-                          lockStateToString(lockState), endpointId);
+            continue;
+        }
 
-            DoorLockServer::Instance().SetLockState(endpointId, lockState, OperationSourceEnum::kRemote, NullNullable, NullNullable,
-                                                    fabricIdx, nodeId);
+        for (int j = 0; j < mUserInStorage.currentCredentialCount; j++)
+        {
+            // If the current credential is a pin type, then check it against pin input. Otherwise ignore
+            if (mCredentials[j].credentialType == CredentialTypeEnum::kPin)
+            {
+                // Read the individual credential at credentialIndex j
+                uint16_t credentialSize = static_cast<uint16_t>(sizeof(LockCredentialInfo));
 
-            return true;
+                chip::StorageKeyName key =
+                    LockCredentialEndpoint(mCredentials[j].credentialIndex, mCredentials[j].credentialType, endpointId);
+
+                error = mStorage->SyncGetKeyValue(key.KeyName(), &mCredentialInStorage, credentialSize);
+
+                if ((error != CHIP_NO_ERROR) && (error != CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND))
+                {
+                    ChipLogError(Zcl, "Error reading credential");
+                    return false;
+                }
+
+                // See if it matches the provided PIN
+                if (mCredentialInStorage.status == DlCredentialStatus::kAvailable)
+                {
+                    continue;
```

Comment:
Might be my personal opinion, but the usage of continue double for loop makes it harder to follow than 
if the follow-up code was just in an if.

Reply:
Inverted this logic and nested the ifs